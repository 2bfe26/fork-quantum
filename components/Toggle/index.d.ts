import { ChangeEventHandler, FC, HTMLAttributes } from 'react';

export interface ToggleProps extends HTMLAttributes<HTMLInputElement>{
  onChange?: ChangeEventHandler<HTMLInputElement>;
  id?: string;
  checked?: boolean;
  theme?: {
    colors?: {};
    spacing?: {};
    baseFontSize?: number;
  };
}

declare const Toggle: FC<ToggleProps>;
export default Toggle;
