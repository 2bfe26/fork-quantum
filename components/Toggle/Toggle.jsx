import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Colors from '../Colors';
import Icon from '../Icon';

const CheckIcon = styled(Icon).attrs({
  name: 'check',
})``;

const CloseIcon = styled(Icon).attrs({
  name: 'close',
})``;

const Wrapper = styled.div`
  display: inline-block;
  position: relative;
  pointer-events: none;
`;

const Switch = styled.div`
  cursor: pointer;
  display: inline-block;
  height: 24px;
  padding-left: 42px;

  &:before {
    background-color: ${Colors.BLACK[400]};
    border-radius: 16px;
    content: '';
    height: 24px;
    position: absolute;
    left: 0;
    top: 0;
    width: 42px;
    transition: background-color 0.2s ease-in-out, color 0.2s ease-in-out,
      transform 0.2s ease-in-out;
  }

  ${CloseIcon},
  ${CheckIcon} {
    align-items: center;
    background-color: ${Colors.BLACK[100]};
    border-radius: 50%;
    display: flex;
    color: ${Colors.BLACK[400]};
    font-size: 12px;
    height: 20px;
    justify-content: center;
    position: absolute;
    left: 2px;
    top: 2px;
    visibility: hidden;
    width: 20px;
    transition: transform 0.2s ease-in-out;
  }

  ${CloseIcon} {
    visibility: visible;
  }
`;

const HiddenCheckbox = styled.input.attrs({
  type: 'checkbox',
  role: 'switch',
})`
  cursor: pointer;
  height: 24px;
  left: 0;
  opacity: 0;
  position: absolute;
  pointer-events: initial;
  width: 100%;

  &:focus + ${Switch}:before {
    box-shadow: 0 2px 6px 0 ${Colors.BLUE[50]};
  }

  &:hover + ${Switch}:before {
    background-color: ${Colors.BLACK[700]};
  }

  &:checked + ${Switch} ${CloseIcon} {
    visibility: hidden;
    transform: translateX(18px);
  }

  &:checked + ${Switch} ${CheckIcon} {
    visibility: visible;
  }

  &:checked + ${Switch}:before {
    background-color: ${Colors.BLUE[500]};
  }

  &:checked:hover + ${Switch}:before {
    background-color: ${Colors.COBALT[500]};
  }

  &:checked + ${Switch} ${CheckIcon} {
    background-color: ${Colors.BLUE[200]};
    color: ${Colors.BLUE[500]};
    transform: translateX(18px);
  }
`;

HiddenCheckbox.displayName = 'HiddenCheckbox';

const Toggle = ({ checked, ...rest }) => (
  <Wrapper>
    <HiddenCheckbox checked={checked} {...rest} />
    <Switch>
      <CloseIcon />
      <CheckIcon />
    </Switch>
  </Wrapper>
);

Toggle.defaultProps = {
  checked: null,
};

Toggle.propTypes = {
  checked: PropTypes.bool,
};

export default Toggle;
