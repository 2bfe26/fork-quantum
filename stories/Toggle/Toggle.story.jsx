import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  Tab,
  Example,
  AutoExample,
  Title,
  StoryContainer,
} from '@catho/quantum-storybook-ui';
import Toggle from '../../components/Toggle';
import { IconFont } from '../../components/GlobalStyle';

const exampleTab = (
  <Tab title="Example">
    <StoryContainer fluid>
      <Title>Left label</Title>
      <Example
        component={
          <div style={{ display: 'flex', alignContent: 'center' }}>
            <label
              htmlFor="left"
              style={{ marginRight: 10, cursor: 'pointer' }}
            >
              Left label with checked true
            </label>
            <Toggle checked id="left" onChange={() => {}} />
          </div>
        }
        code={`<div style={{ display: 'flex', alignContent: 'center' }}>
  <label htmlFor="left" style={{ marginRight: 10 }}>Left label</label>
  <Toggle checked id="left" />
</div>`}
      />
      <Title>Right label</Title>
      <Example
        component={
          <div style={{ display: 'flex', alignContent: 'center' }}>
            <Toggle id="right" />
            <label
              htmlFor="right"
              style={{ marginLeft: 10, cursor: 'pointer' }}
            >
              Right label without checked prop
            </label>
          </div>
        }
        code={`<div style={{ display: 'flex', alignContent: 'center' }}>
  <Toggle id="right" />
  <label htmlFor="right" style={{ marginLeft: 10 }}>Right label</label>
</div>`}
      />
    </StoryContainer>
  </Tab>
);

const description = `Toggle is a control that is used to quickly switch
between two possible states.`;

storiesOf('Forms', module).add('Toggle', () => (
  <>
    <IconFont />
    <AutoExample
      description={description}
      component={Toggle}
      additionalTabs={exampleTab}
    />
  </>
));
