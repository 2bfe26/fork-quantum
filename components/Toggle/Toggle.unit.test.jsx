import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Toggle from './Toggle';
import 'jest-styled-components';

describe('<Toggle /> ', () => {
  describe('Snapshots', () => {
    it('should match snapshot without props', () => {
      expect(renderer.create(<Toggle />).toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with checked prop', () => {
      expect(renderer.create(<Toggle checked />).toJSON()).toMatchSnapshot();
    });
  });

  describe('Correctly props', () => {
    it('should pass onChange prop to input element', () => {
      const onChangeMock = () => {};
      const toggle = shallow(<Toggle onChange={onChangeMock} />);
      const checkbox = toggle.find('HiddenCheckbox');

      expect(checkbox.props().onChange).toBe(onChangeMock);
    });

    it('should pass checked prop to input element', () => {
      const toggle = shallow(<Toggle checked />);
      const checkbox = toggle.find('HiddenCheckbox');

      expect(checkbox.props().checked).toBe(true);
    });
  });
});
