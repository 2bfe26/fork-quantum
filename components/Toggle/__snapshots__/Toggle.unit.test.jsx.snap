// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component  should match the snapshot 1`] = `
<div
  className="sc-htpNat hxsEuH"
>
  <label
    checked={false}
    className="sc-ifAKCX cSrAET sc-bwzfXH eQhPVJ"
    disabled={false}
  >
    <input
      checked={false}
      className="sc-bxivhb cRvGhf"
      disabled={false}
      id=""
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
    
  </label>
</div>
`;
